wa/workloads/speedometer/__init__.py

```diff
@@ -12,57 +12,365 @@
+                    "Speedometer Score", score, "Runs per minute", lower_is_better=False
+                )
+                score_read = True
+            else:
+                if iterations >= 10:
+                    raise WorkloadError(
+                        "The Speedometer workload has failed. No score was obtainable."
+                    )
+                else:
+                    # Sleep and retry...
+                    time.sleep(2)
+                    iterations += 1
+
+    def teardown(self, context):
+        super(Speedometer, self).teardown(context)
+
+        # The browser's processes can stick around and have minor impact on
+        # other performance sensitive workloads, so make sure we clean up.
+        self.target.execute("am force-stop {}".format(self.chrome_package))
+
+        if self.cleanup_assets:
+            # The only thing left on device was the UI dump created by uiautomator.
+            self.target.execute("rm {}".format(self.ui_dump_loc), as_root=True)
+
+    @once
+    def finalize(self, context):
+        super(Speedometer, self).finalize(context)
+
+        # Shutdown the locally hosted version of Speedometer
+        self.archive_server.stop(self.target)
+
+
+class ArchiveServerThread(threading.Thread):
+    """Thread for running the HTTPServer"""
+
+    def __init__(self, httpd):
+        self._httpd = httpd
+        threading.Thread.__init__(self)
+
+    def run(self):
+        self._httpd.serve_forever()
+
+
+class DifferentDirectoryHTTPRequestHandler(SimpleHTTPRequestHandler):
+    """A version of SimpleHTTPRequestHandler that allows us to serve
+    relative files from a different directory than the current one.
+    This directory is captured in |document_root|. It also suppresses
+    logging."""
+
+    def translate_path(self, path):
+        document_root = self.server.document_root
+        path = SimpleHTTPRequestHandler.translate_path(self, path)
+        requested_uri = os.path.relpath(path, os.getcwd())
+        return os.path.join(document_root, requested_uri)
+
+    # Disable the logging.
+    # pylint: disable=redefined-builtin
+    def log_message(self, format, *args):
+        pass
+
+
+class ArchiveServer(object):
+    def __init__(self):
+        self._port = None
+
+    def start(self, document_root, target):
+        # Create the server, and find out the port we've been assigned...
+        self._httpd = HTTPServer(("", 0), DifferentDirectoryHTTPRequestHandler)
+        # (This property is expected to be read by the
+        #  DifferentDirectoryHTTPRequestHandler.translate_path method.)
+        self._httpd.document_root = document_root
+        _, self._port = self._httpd.server_address
+
+        self._thread = ArchiveServerThread(self._httpd)
+        self._thread.start()
+
+        adb_command(target.adb_name, "reverse tcp:{0} tcp:{0}".format(self._port))
+
+    def stop(self, target):
+        adb_command(target.adb_name, "reverse --remove tcp:{}".format(self._port))
```

Comment:
As discussed, it seems this is an issue with devices before Android 9 when connected via TCP. Added documentation to note which Android versions are supported (9+ via TCP, 5+ via USB).

Reply:
What is `adb_server` set to? Sorry I'm not familiar with the usage here - is that something that can be found in `self`, i.e. `adb_server=self.adb_server` if you've set an ADB server in your config?

Reply:
Ah sorry, it would set similar to they way you retrieve the `adb_name` ie. `adb_server=target.adb_server` as it would be a part of the supplied device config.

Reply:
As discussed elsewhere, this isn't going to work without some kind of extra port forwarding being setup between the machine running WA and the machine running the remote ADB server - so we detect if adb_server is non-None in target, and refuse to run the workload in this case, saying it isn't supported.